compiler/x/codegen/BinaryCommutativeAnalyser.cpp

```diff
@@ -860,30 +860,17 @@ TR::Register *TR_X86BinaryCommutativeAnalyser::integerAddAnalyserImpl(TR::Node
       }
    else if (getCopyRegs())
       {
-      TR::Register *tempReg;
+      TR::Register* tempReg = _cg->allocateRegister();
       if (firstRegister->containsCollectedReference() ||
           secondRegister->containsCollectedReference() ||
           firstRegister->containsInternalPointer() ||
           secondRegister->containsInternalPointer())
          {
-         if (root->isInternalPointer())
+         if (root->isInternalPointer() && root->getPinningArrayPointer())
             {
-            tempReg = _cg->allocateRegister();
-            if (root->getPinningArrayPointer())
-               {
-               tempReg->setContainsInternalPointer();
-               tempReg->setPinningArrayPointer(root->getPinningArrayPointer());
-               }
+            tempReg->setContainsInternalPointer();
+            tempReg->setPinningArrayPointer(root->getPinningArrayPointer());
             }
-         else if (comp->generateArraylets() && root->getOpCodeValue() == TR::aiadd)
-            // arraylets: aiadd is technically internal pointer into spine object, but isn't marked as internal pointer
-            tempReg = _cg->allocateRegister();
-         else
```

Comment:
I presume this allocateCollectedReferenceRegister section of code is the one that was causing problems ? 
Could you please elaborate on the actual problem that you are fixing ?

Reply:
The result register (`tempReg` here) was incorrectly marked as a collected reference, when the `aiadd`/`aladd` node is not internal pointer. As a result, an invalid GC map entry may be introduced and causes GC assertions.
I cannot think of any valid reasons that this result could be collected; therefore, all branches are now to allocate a "non-collected" register instead of "collected" ones.

Reply:
It seems we would have used an (uncollected) register if the isInternalPointer flag would have been "true" on this aiadd/aladd. I expect this would be set on all aiadd/aladd in Java but I guess you saw some case where that node flag was not set (?) If so, how did such a node originate ?

Reply:
It is from Value Propagation where a `TR::arraycopy` gets generated. The source and destination children are `TR::aladd`.

Reply:
If these aladds created by arraycopy generation in value propagation don't have the internal pointer flag set, then we should consider setting the flag and see if that fixes the failure.

Reply:
But either way `aladd` shouldn't return a collected reference. It doesn't make sense to create a collected reference from [address+offset].

Reply:
Agreed, but it's preferable to fix a bug in common code if we can so that each platform does'nt have to make changes.

Reply:
Good suggestion. @liqunl kindly agreed to investigate why `root->isInternalPointer()` answered no as a separate issue.

Reply:
It seems like VP arraycopy transformation creates an aiadd/aladd without setting the flag. I'm trying to find where the node is created.